cpp/dolfinx/mesh/MeshTags.h

```diff
@@ -0,0 +1,151 @@
+// Copyright (C) 2020 Michal Habera
+//
+// This file is part of DOLFINX (https://www.fenicsproject.org)
+//
+// SPDX-License-Identifier:    LGPL-3.0-or-later
+
+#pragma once
+
+#include "Geometry.h"
+#include "Mesh.h"
+#include "Topology.h"
+#include <dolfinx/common/IndexMap.h>
+#include <dolfinx/common/UniqueIdGenerator.h>
+#include <map>
+#include <memory>
+#include <utility>
+
+namespace dolfinx
+{
+namespace mesh
+{
+
+/// A MeshTags is a class used to tag mesh entities using their
+/// local-to-process index and an attached value.
+/// MeshTags is a sparse data storage class, since it allows to
+/// tag only few mesh entities. This class sorts and removes duplicates
+/// in indices on construction.
+/// @tparam Type
+template <typename T>
+class MeshTags
+{
+public:
+  /// Create from entities of given dimension on a mesh
+  /// @param[in] mesh The mesh associated with the tags
+  /// @param[in] dim Topological dimension of mesh entities
+  ///    to tag.
+  /// @param[in] indices Array of indices, will be copied, sorted
+  ///    with duplicates removed. Local-to-process.
+  /// @param[in] values Array of values attached to indices,
+  ///    will be copied, sorted and duplicates removed according
+  ///    to indices array.
+  MeshTags(const std::shared_ptr<const Mesh>& mesh, int dim,
```

Comment:
It would be helpful to add a flag that allows caller to say that the input data is unique and sorted.